# RPT1

Reporters are either a `Transformer` for a `TestsStream` or a generator function. Test events pass through the reporter and are output based on the reporter's code.

In this case, the reporter simply prints the event name and the `event` object so we can better understand test flow and what data is available for each event in the stream.

```
TEST:ENQUEUE
{
   "type": "test:enqueue",
   "data": {
      "nesting": 0,
      "name": "/workspace/test/test.spec.ts"
   }
}

TEST:DEQUEUE
{
   "type": "test:dequeue",
   "data": {
      "nesting": 0,
      "name": "/workspace/test/test.spec.ts"
   }
}

TEST:STDERR
{
   "type": "test:stderr",
   "data": {
      "file": "/workspace/test/test.spec.ts",
      "message": "(node:9307) ExperimentalWarning: `--experimental-loader` may be removed in the future; instead use `register()`:\n"
   }
}

TEST:STDERR
{
   "type": "test:stderr",
   "data": {
      "file": "/workspace/test/test.spec.ts",
      "message": "--import 'data:text/javascript,import { register } from \"node:module\"; import { pathToFileURL } from \"node:url\"; register(\"ts-node/esm\", pathToFileURL(\"./\"));'\n"
   }
}

TEST:STDERR
{
   "type": "test:stderr",
   "data": {
      "file": "/workspace/test/test.spec.ts",
      "message": "(Use `node --trace-warnings ...` to show where the warning was created)\n"
   }
}

TEST:STDOUT
{
   "type": "test:stdout",
   "data": {
      "file": "/workspace/test/test.spec.ts",
      "message": "PASS stdout TEST-PASS\n"
   }
}

TEST:STDERR
{
   "type": "test:stderr",
   "data": {
      "file": "/workspace/test/test.spec.ts",
      "message": "FAIL stderr TEST-FAIL\n"
   }
}

TEST:ENQUEUE
{
   "type": "test:enqueue",
   "data": {
      "nesting": 0,
      "name": "jestExpect",
      "line": 3,
      "column": 1,
      "file": "file:///workspace/test/test.spec.ts"
   }
}

TEST:DEQUEUE
{
   "type": "test:dequeue",
   "data": {
      "nesting": 0,
      "name": "jestExpect",
      "line": 3,
      "column": 1,
      "file": "file:///workspace/test/test.spec.ts"
   }
}

TEST:ENQUEUE
{
   "type": "test:enqueue",
   "data": {
      "nesting": 1,
      "name": "TEST-PASS",
      "line": 4,
      "column": 5,
      "file": "file:///workspace/test/test.spec.ts"
   }
}

TEST:DEQUEUE
{
   "type": "test:dequeue",
   "data": {
      "nesting": 1,
      "name": "TEST-PASS",
      "line": 4,
      "column": 5,
      "file": "file:///workspace/test/test.spec.ts"
   }
}

TEST:ENQUEUE
{
   "type": "test:enqueue",
   "data": {
      "nesting": 1,
      "name": "TEST-FAIL",
      "line": 8,
      "column": 5,
      "file": "file:///workspace/test/test.spec.ts"
   }
}

TEST:START
{
   "type": "test:start",
   "data": {
      "nesting": 0,
      "name": "jestExpect",
      "line": 3,
      "column": 1,
      "file": "file:///workspace/test/test.spec.ts"
   }
}

TEST:START
{
   "type": "test:start",
   "data": {
      "nesting": 1,
      "name": "TEST-PASS",
      "line": 4,
      "column": 5,
      "file": "file:///workspace/test/test.spec.ts"
   }
}

TEST:PASS
{
   "type": "test:pass",
   "data": {
      "name": "TEST-PASS",
      "nesting": 1,
      "testNumber": 1,
      "details": {
         "duration_ms": 1.95771
      },
      "line": 4,
      "column": 5,
      "file": "file:///workspace/test/test.spec.ts"
   }
}

TEST:DEQUEUE
{
   "type": "test:dequeue",
   "data": {
      "nesting": 1,
      "name": "TEST-FAIL",
      "line": 8,
      "column": 5,
      "file": "file:///workspace/test/test.spec.ts"
   }
}

TEST:START
{
   "type": "test:start",
   "data": {
      "nesting": 1,
      "name": "TEST-FAIL",
      "line": 8,
      "column": 5,
      "file": "file:///workspace/test/test.spec.ts"
   }
}

TEST:FAIL
{
   "type": "test:fail",
   "data": {
      "name": "TEST-FAIL",
      "nesting": 1,
      "testNumber": 2,
      "details": {
         "duration_ms": 2.176413,
         "error": {
            "failureType": "testCodeFailure",
            "cause": {
               "matcherResult": {
                  "message": "expect(received).toBeLessThan(expected)\n\nExpected: < 4\nReceived:   5",
                  "pass": false
               }
            },
            "code": "ERR_TEST_FAILURE"
         }
      },
      "line": 8,
      "column": 5,
      "file": "file:///workspace/test/test.spec.ts"
   }
}

TEST:PLAN
{
   "type": "test:plan",
   "data": {
      "nesting": 1,
      "count": 2,
      "line": 3,
      "column": 1,
      "file": "file:///workspace/test/test.spec.ts"
   }
}

TEST:FAIL
{
   "type": "test:fail",
   "data": {
      "name": "jestExpect",
      "nesting": 0,
      "testNumber": 1,
      "details": {
         "duration_ms": 6.346492,
         "type": "suite",
         "error": {
            "failureType": "subtestsFailed",
            "cause": "1 subtest failed",
            "code": "ERR_TEST_FAILURE"
         }
      },
      "line": 3,
      "column": 1,
      "file": "file:///workspace/test/test.spec.ts"
   }
}

TEST:PLAN
{
   "type": "test:plan",
   "data": {
      "nesting": 0,
      "count": 1
   }
}

TEST:DIAGNOSTIC
{
   "type": "test:diagnostic",
   "data": {
      "nesting": 0,
      "message": "tests 2"
   }
}

TEST:DIAGNOSTIC
{
   "type": "test:diagnostic",
   "data": {
      "nesting": 0,
      "message": "suites 1"
   }
}

TEST:DIAGNOSTIC
{
   "type": "test:diagnostic",
   "data": {
      "nesting": 0,
      "message": "pass 1"
   }
}

TEST:DIAGNOSTIC
{
   "type": "test:diagnostic",
   "data": {
      "nesting": 0,
      "message": "fail 1"
   }
}

TEST:DIAGNOSTIC
{
   "type": "test:diagnostic",
   "data": {
      "nesting": 0,
      "message": "cancelled 0"
   }
}

TEST:DIAGNOSTIC
{
   "type": "test:diagnostic",
   "data": {
      "nesting": 0,
      "message": "skipped 0"
   }
}

TEST:DIAGNOSTIC
{
   "type": "test:diagnostic",
   "data": {
      "nesting": 0,
      "message": "todo 0"
   }
}

TEST:DIAGNOSTIC
{
   "type": "test:diagnostic",
   "data": {
      "nesting": 0,
      "message": "duration_ms 3882.949864"
   }
}
```

## test:stdout and test:stderr

These events log some output at the beginning of the test but they also log `console.log()` and `console.error()` output from the tests themselves. I don't see a reliable way to distinguish test output from Node output. It's interesting to note that these lines appear before their tests' enqueue/dequeue events but include data from the test scope (`t.name`).

```
TEST:STDOUT
{
   "type": "test:stdout",
   "data": {
      "file": "/workspace/test/test.spec.ts",
      "message": "PASS stdout TEST-PASS\n"
   }
}

TEST:STDERR
{
   "type": "test:stderr",
   "data": {
      "file": "/workspace/test/test.spec.ts",
      "message": "FAIL stderr TEST-FAIL\n"
   }
}
```

## test:enqueue and test:dequeue

The test file is enqueued and dequeued, then node warnings, then stdout/stderr. Then the first test (nesting level 0) is enqueued and dequeued. It includes two nested tests (nesting level 1). The first L1 test is enqueued and dequeued and the second L1 test is enqueued. Then execution of the L0 test begins which starts the first L1 test. When the first L1 test finishes (pass), the second L1 test is dequeued and started. The second L1 test fails, which causes the L0 test to fail.

The early L1 dequeue seems strange, but the rest makes sense to me.

My conclusion is, I'm not very interested in enqueue/dequeue unless it helps me track something meaningful about what should be reported. For example, should I really report the L0 test as failed too? Or should I ignore it?

But `event.data.details.error.failureType` tells me if the test itself failed (`testCodeFailure`) or a subtest failed (`subtestsFailed`), so probably not important for my goals.

```
TEST:FAIL
{
   "type": "test:fail",
   "data": {
      "name": "jestExpect",
      "nesting": 0,
      "testNumber": 1,
      "details": {
         "duration_ms": 6.346492,
         "type": "suite",
         "error": {
            "failureType": "subtestsFailed",
            "cause": "1 subtest failed",
            "code": "ERR_TEST_FAILURE"
         }
      },
      "line": 3,
      "column": 1,
      "file": "file:///workspace/test/test.spec.ts"
   }
}
```

## test:pass and test:fail

These two events are important. They identify the test in question and what happened with it. The samples below are for lowest-level test failures, while the failure above is for a parent test with failed subtests.

```
TEST:PASS
{
   "type": "test:pass",
   "data": {
      "name": "TEST-PASS",
      "nesting": 1,
      "testNumber": 1,
      "details": {
         "duration_ms": 1.95771
      },
      "line": 4,
      "column": 5,
      "file": "file:///workspace/test/test.spec.ts"
   }
}
TEST:FAIL
{
   "type": "test:fail",
   "data": {
      "name": "TEST-FAIL",
      "nesting": 1,
      "testNumber": 2,
      "details": {
         "duration_ms": 2.176413,
         "error": {
            "failureType": "testCodeFailure",
            "cause": {
               "matcherResult": {
                  "message": "expect(received).toBeLessThan(expected)\n\nExpected: < 4\nReceived:   5",
                  "pass": false
               }
            },
            "code": "ERR_TEST_FAILURE"
         }
      },
      "line": 8,
      "column": 5,
      "file": "file:///workspace/test/test.spec.ts"
   }
}
```

We can get the test's name, filename, line and column in the file, and how long it ran. Nesting could be used to indent tests below parent tests (number of indents to apply).

So, it should be possible to format lines like below with some color coding:

PASS TEST-PASS in 1.96ms (test/test.spec.ts:4:5)
FAIL TEST-FAIL in 2.18ms (test/test.spec.ts:8:5)

I also know `event.data.details.cause.matcherResult.message` usually includes color coding like below, so printed as a string, it should give nice output for errors.

`\u001b[2mexpect(\u001b[22m\u001b[31mreceived\u001b[39m\u001b[2m).\u001b[22mtoBeLessThan\u001b[2m(\u001b[22m\u001b[32mexpected\u001b[39m\u001b[2m)\u001b[22m\n\nExpected: < \u001b[32m4\u001b[39m\nReceived:   \u001b[31m5\u001b[39m`

I don't see anything that identifies the specific line in the test file that failed. In fact, the line numbers are wrong, which is curious because the default reporter's stack dump starts with `at TestContext.<anonymous> (/workspace/test/test.spec.ts:12:19)`, which identifies the correct line.

(It's late. I'll look at this more tomorrow.)